File: strings/1249_minimum_remove_to_make_valid_parentheses.ts
function minRemoveToMakeValid(s: string): string {

    let ends = 0, opened = 0;
    let ans = "";

    for(let l of s){
        if(l === ")"){
            ends ++;
        }
    }

    for(let l of s){

        if(l === "("){
            if(ends < 1){
                continue
            } else {
                ends -= 1;
                opened += 1;
            }
        }
        if(l === ")"){
            if(opened < 1){
                ends -= 1
                continue
            } else {
                opened -= 1;
            }
        }

        ans += l;
    }

    return ans;

};

File: strings/1249_minimum_remove_to_make_valid_parentheses.rs
impl Solution {
    pub fn min_remove_to_make_valid(s: String) -> String {

        let mut ends:i32 = 0;
        let mut opened:i32 = 0;
        let mut ans:String = String::new();

        for c in s.chars() {
            if c == ')' {
                ends += 1;
            }
        }

        for c in s.chars() {

            if c == '(' {
                if ends < 1 {
                    continue
                } else {
                    ends -= 1;
                    opened += 1;
                }
            }
            if c == ')' {
                if opened < 1 {
                    ends -= 1;
                    continue
                } else {
                    opened -= 1;
                }
            }

            ans.push(c);
        }

        ans

    }
}